# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [4]:
import math
def prvocisla(n):
    def test_prvocisla(j):
        test = True
        i = 2
        while i <= math.sqrt(j):
            if j % i == 0:
                test =False
                break
            else: 
                i += 1
        return test
    seznam_prvocisel = []
    k = 2
    while  len(seznam_prvocisel) < n:
        if test_prvocisla(k):
            seznam_prvocisel.append(k)
        k += 1
    return seznam_prvocisel


In [5]:
prvocisla(10)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]


# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [4]:
%%writefile modul_ze_tri_funkci.py
# Výpočet druhé odmocniny pomocí prostých iterací
def druha_odmocnina(cislo, pocet_iteraci):
    hodnota = cislo #odmocnovane cislo jako nulty clen iteracni posloupnosti
    for j in range(pocet_iteraci): #cyklus zajistujici vypocet odmocniny jako pevneho bodu kontraktivniho zobrazeni
        hodnota = (cislo/hodnota+hodnota)/2
    return hodnota
# Výpočet obsahu vepsaných úhelníků
def obsaheuler(pocet_iteraci):
    strana = 1  
    vyska =  druha_odmocnina(3, 100)/2 #vyska v sestiuhelniku
    for j in range(pocet_iteraci):
        strana = druha_odmocnina((strana/2)**2+(1-vyska)**2, 100) #vypocet zakladny
        vyska =  druha_odmocnina(1-(strana/2)**2, 100)
    return 6*(2**pocet_iteraci)*strana*vyska/2 #funkce vraci obsah vepsaneho 6*2**pocet_iteraci uhelnika
# Součet řady vedoucí k aproximaci pí
def obsahnewton(pocet_clenu):
    a_i=1/16 #stanoveni prvniho clenu
    soucet=(1/3)*a_i #soucet pro i=1
    for i in range(2,pocet_clenu+1):  
        a_i=(1/4)*a_i*(2*i-3)/(2*i) #vypocet nasledujiciho koeficientu
        soucet += (1/(2*i+1))*a_i   #dalsi scitanec k souctu rady
    return 12*(1/2-druha_odmocnina(3,1000)/8-soucet) #funkce vraci aproximaci cisla pi pomoci n clenu rady

Overwriting modul_ze_tri_funkci.py


In [6]:
import modul_ze_tri_funkci as mtrif
%time mtrif.druha_odmocnina(2,8)
%timeit mtrif.druha_odmocnina(2,8)

CPU times: total: 0 ns
Wall time: 0 ns
2.19 µs ± 385 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [7]:
%%time
mtrif.obsaheuler(50)
mtrif.obsahnewton(50)

CPU times: total: 0 ns
Wall time: 4.07 ms


3.1415926535897936

In [8]:
%%timeit
mtrif.obsaheuler(50)
mtrif.obsahnewton(50)

2.22 ms ± 1 ms per loop (mean ± std. dev. of 7 runs, 1,000 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [9]:
n=5 # Výpis odhadu pí pomocí obsahů n-úhelníků
for m in range(2,n+1):
    j = True
    l = 1 
    while str(mtrif.obsaheuler(m))[:l] == str(math.pi)[:l] :
        l += 1
    print(f"n={m}, pi={mtrif.obsaheuler(m):.16f}, pocet správných cifer={l-2}")


n=2, pi=3.1058285412302484, pocet správných cifer=2
n=3, pi=3.1326286132812378, pocet správných cifer=2
n=4, pi=3.1393502030468676, pocet správných cifer=2
n=5, pi=3.1410319508905102, pocet správných cifer=4


In [10]:
n=9 # Výpis odhadu pí pomocí řady
for m in range(2,n+1):
    j = True
    l = 1 
    while str(mtrif.obsahnewton(m))[:l] == str(math.pi)[:l] :
          l += 1 
    #print(f"n={n}, pi={mtrif.obsaheuler(n):.10f}, pocet správných cifer={1+abs(int(math.log10((abs(mtrif.obsaheuler(n)-math.pi)))))}")
    print(f"n={m}, pi={mtrif.obsahnewton(m):.16f}, pocet správných cifer={l-2}")

n=2, pi=3.1425487886466845, pocet správných cifer=3
n=3, pi=3.1417117350752557, pocet správných cifer=4
n=4, pi=3.1416100098148387, pocet správných cifer=4
n=5, pi=3.1415954446070975, pocet správných cifer=6
n=6, pi=3.1415931337808694, pocet správných cifer=6
n=7, pi=3.1415927403902137, pocet správných cifer=7
n=8, pi=3.1415926698836163, pocet správných cifer=8
n=9, pi=3.1415926567409396, pocet správných cifer=9



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [11]:
def generator_fibonacci(n):
    for j in range(n):
        yield j+(j+1)

In [12]:
n = 8
fibonacci_gen = generator_fibonacci(n)
for j in range(n):
    print(next(fibonacci_gen))

1
3
5
7
9
11
13
15



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [13]:
s1 = [i for i in range(1,11)]
s2 = [j**2 for j in range(1,11)]
s3 = [k for k in range(2,101) if all([False for j in range(2,k-1) if k % j == 0])]


In [14]:
print(s1)
print(s2)
print(s3)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [15]:
%%writefile dutext.txt
Soubor s nějakým textem, co přijde uložit.
Další a poslední řádek.

Overwriting dutext.txt


In [16]:
ctu = open("dutext.txt","r")
print(ctu.read())
ctu.close()

Soubor s nějakým textem, co přijde uložit.
Další a poslední řádek.



In [17]:
with open("pro_zapis.txt", "w", newline="") as duzapis:
    duzapis.write("Vkládám textový řetězec \n" "o dvou řádcích")
    duzapis.close()
    

In [18]:
ctu = open("pro_zapis.txt","r")
print(ctu.read())
ctu.close()

Vkládám textový řetězec 
o dvou řádcích


In [19]:
with open("tabulka.txt", "w", newline="") as duzapis:
    for i in range(1, 11):
        duzapis.write(f"{i:>4}, {i*i:>4},   {str(math.sqrt(i))[:8]} \n")
    duzapis.close()
    
    

In [20]:
ctu = open("tabulka.txt","r")
print(ctu.read())
ctu.close()

   1,    1,   1.0 
   2,    4,   1.414213 
   3,    9,   1.732050 
   4,   16,   2.0 
   5,   25,   2.236067 
   6,   36,   2.449489 
   7,   49,   2.645751 
   8,   64,   2.828427 
   9,   81,   3.0 
  10,  100,   3.162277 



# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [27]:
class Operator:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        return Addition(self,other) #TODO(použijte Addition)
        
    
    def __sub__(self, other):
        return Addition(self,Multiplication(cst(-1),other)) # TODO (použijte Addition a Multiplication se zápornou konstantu)
        
    def __mul__(self, other):
        return Multiplication(self, other) # TODO (použijte Multiplication)
       
    
class BinaryOperator(Operator):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Operator):
    def __init__(self, argument):
            self.argument = argument
    
    def __str__(self):
        return type(self).__name__ + "(" + str(self.argument) + ")"
    
class Addition(BinaryOperator):
    def __str__(self):
        #return "(" + str(self.levy_argument) + "+" +str(self.pravy_argument) +")" # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        return  str(self.levy_argument)  + "+"  + str(self.pravy_argument)  # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        
    def diff(self):
        return Addition(self.levy_argument.diff(),self.pravy_argument.diff())# TODO
        #pass

class Multiplication(BinaryOperator):
    def __str__(self):
        return "(" + str(self.levy_argument) + ")" + "*" + "(" + str(self.pravy_argument) + ")"  # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
    
    def diff(self):
        return Addition(Multiplication(self.levy_argument.diff(),self.pravy_argument),Multiplication(self.levy_argument,self.pravy_argument.diff()))# TODO
       
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    
    def __str__(self):
        return  str(self.value)  # TODO vypíšeme něco jako value
       
    def diff(self):
        return cst(0)# TODO
        
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        return "x" # TODO vypíšeme něco jako x   
        
    def diff(self):
        return cst(1)     # TODO
               
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        return  str(self.argument) + "^" + "(" + str(self.exponent) + ")"# !!! Tady jsem přidal závorky, vypadalo by asi blbě x^-2 ....Lépe: x^(-2) !!!# TODO vypíšeme něco jako argument^exponent
        
    def diff(self):
        return Multiplication(Multiplication(self.exponent,mocnina(self.argument,self.exponent-1)),self.argument.diff()) # TODO
     
class sin(UnaryOperator):
    def diff(self):
        return Multiplication(cos(self.argument),self.argument.diff()) # TODO
        
class cos(UnaryOperator):
    def diff(self):
        return Multiplication(Multiplication(cst(-1),sin(self.argument)),self.argument.diff()) # TODO
        
class exp(UnaryOperator):
    def diff(self):
        return Multiplication(exp(self.argument),self.argument.diff())
        # TODO        

class ln(UnaryOperator):
    def diff(self):
        return Multiplication(mocnina(self.argument,cst(-1)),self.argument.diff())
    # TODO        

class frc(UnaryOperator):
    def __str__(self):
        return "1/(" + str(self.argument) + ")"
    def diff(self):
        return Multiplication(Multiplication(cst(-1),mocnina(self.argument,cst(-2))),self.argument.diff())
        # TODO         
        


In [24]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
f1 = sin(x + cos(x*x)) + cst(1)
print(f1)
print(f1.diff())

sin(x+cos((x)*(x)))+1
(cos(x+cos((x)*(x))))*(1+((-1)*(sin((x)*(x))))*((1)*(x)+(x)*(1)))+0


můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [25]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


x^(2)+(2)*(x)+1+(2)*(x^(2))
((2)*(x^(1)))*(1)+(0)*(x)+(2)*(1)+0+(0)*(x^(2))+(2)*(((2)*(x^(1)))*(1))


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [26]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


x^(2)+(2)*(x)+exp(x)
((2)*(x^(1)))*(1)+(0)*(x)+(2)*(1)+(exp(x))*(1)


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


In [28]:
f4 = frc(sin(x))
print(f4)
print(f4.diff())

1/(sin(x))
((-1)*(sin(x)^(-2)))*((cos(x))*(1))


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...